docs/hugo/content/design/secrets.md

```diff
@@ -0,0 +1,669 @@
+
+## Proposal
+
+### User specified secrets
+
+User specified secrets will be detected and transformed from `string` to a `SecretRef`:
+
+```go
+type SecretRef struct {
+    // SecretName is the name of the secret. The secret must be in the same namespace as the resource.
+    SecretName string
+
+    // Key is the key in the secret to use.
+    Key string
+}
+```
+
+Detection will be done with a combination of:
+1. Additions to ASO's configuration file to flag particular fields as secret.
+2. Using the `"format": "password"` data from Swagger, such as that used by 
+   [MySQL Flexible Server](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/mysql/resource-manager/Microsoft.DBforMySQL/stable/2021-05-01/mysql.json#L1662). 
+   Note that not all specs have this, for example
+   [VMSS does not](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json#L9397).
+3. Using the `x-ms-secret` annotation, which I assume has the same meaning as `"format": "password"` although it's not actually exactly documented anywhere.
+
+This is a place where we can push changes upstream to flag things as passwords if they're not being flagged.
+
+#### Lifecycle
+Since these secrets were created _by the user_, the user owns the lifecycle of these secrets. They could be using the same secret across many resources, or 
+intending to use this secret on a resource they have not created yet. As such, the lifecycle of these secrets must be controlled by the user.
+
+Rollover will be supported by triggering events on the associated resource when the secret is modified. Since multiple custom resources might be using 
+the same secret, this could trigger reconciles on multiple resources. An existing pattern has been established for this in 
+[mysqlserver_controller.go](https://github.com/Azure/azure-service-operator/blob/main/controllers/mysqlserver_controller.go#L51) of ASO v1.
+
+#### Open questions
+
+**Should we support KeyVault inputs?**
+
+I think we can get away without it at least initially... but maybe we need to fix up the shape of the `SecretRef` a bit to more easily support it in the future?
+
+```yaml
+  ...
+  secret:
+    kubernetes:
+      name: foo
+      key: bar
+    keyVault:
+      url: ??
+```
+
+Alternatively, we can keep this focused on Kubernetes secrets and if we want to expand to KeyVault secrets in the future just that _that_ on as a separate property
+so that the base (Kubernetes) case is always clean:
+```yaml
+  secret:
+    name: foo
+    key: bar
+    keyVault:
+      url: ??
+```
+
+<!-- TODO: Is there specific API guidance from Kubernetes on how to model these unions? There's some stuff discussed here, but it doesn't seem widely adopted: https://github.com/ecordell/enhancements-1/blob/master/keps/sig-api-machinery/20190325-unions.md -->
+
+Conclusion: TBD
```

Comment:
Yeah that's (basically) what Kubernetes suggests I _think_ based on the linked design. I am not actually sure how we'll it's modelled in the backing Go types though. In other REST APIs I've seen the actual SDKs/types end up with an abstract base type with a single property (the `secretType` in this case) and then derived types have other properties. This results in having to perform casts to actually get at the more concrete derived type (at least during GETs).

Reply:
and the usually suggested alternative I've seen for fixing that problem is to instead of doing an inheritance approach do a `oneOf` with a type that looks like:
```
type Example struct {
    optionA *TypeA
    optionB *TypeB
}
```

Reply:
Upon further reading of their spec, actually they suggest:
```go
type Example struct {
    discriminator DiscriminatorEnum
    optionA *TypeA
    optionB *TypeB
}
```

Reply:
I've updated this question with more notes about what Kubernetes recommends. They actually allow for either of the two proposed shapes here, so we will need to decide which one we like more.